### PR TITLE
Add resolve files

### LINKED ISSUE #301: Need  resolve_files() method
### Description

Datachain cannot enforce immutability of storage, so it helps to check if the files are still there.
Currently there is no way to verify, and a chain will just crash when missing a file.

Suggestion is to introduce a method 

```
resolve_files(signal = "file")
```

Which checks file object under "signal" name, and marks File object as valid or invalid.

This requires introduction of a new field of the type Boolean or Date which will mark validity as True/False or None/"last accessed" fashion.

### PR DESCRIPTION
**Pull request by EdwardLi-coder**
Fix: Need resolve_files() method #301

Closes #301

### CODE REVIEW

**Comment by EdwardLi-coder** [comment]
I've attempted to resolve this issue, but my test is still failing. I'd like to understand the reason why the 'is_valid' field is not being successfully updated. Could you please help me identify the cause of this problem? Thank you."
# test_resolve_files

```python
def test_resolve_files(test_session):
    valid_file = File(path="valid.txt", size=10)
    invalid_file = File(path="invalid.txt", size=5)

    dc = DataChain.from_values(file=[valid_file, invalid_file], session=test_session)

    def mock_open(self):
        if self.path == "valid.txt":
            return MagicMock()
        raise OSError("File not found")

    with patch("datachain.lib.file.File.open", mock_open):
        resolved_dc = dc.resolve_files()

        results = list(resolved_dc.collect("file"))
        print("results: ",results)
        assert len(results) == 2
        assert results[0].is_valid
AssertionError: assert None

where None = File(source='', path='valid.txt', size=10, version='', etag='', is_latest=True, last_modified=datetime.datetime(1970, 1, 1, 0, 0, tzinfo=datetime.timezone.utc), location=None, vtype='', is_valid=None).is_valid
/Users/lifei/PycharmProjects/datachain/tests/unit/lib/test_datachain.py:1633: AssertionError

**Comment by mattseddon** [comment]
@EdwardLi-coder to help you fully understand what is going on underneath you can set the `DEBUG_SHOW_SQL_QUERIES=1` environment variable to see the SQL generated by DataChain.

From the following, it looks like it is not possible to update a field under an existing Feature right now as the result will not be included in the final output:

```sql
    SELECT
        "udf_bL9vdT".sys__id,
        "udf_bL9vdT".sys__rand,
        "udf_bL9vdT".file__source,
        "udf_bL9vdT".file__path,
        "udf_bL9vdT".file__size,
        "udf_bL9vdT".file__version,
        "udf_bL9vdT".file__etag,
        "udf_bL9vdT".file__is_latest,
        "udf_bL9vdT".file__last_modified,
        "udf_bL9vdT".file__location,
        "udf_bL9vdT".file__vtype,
        "udf_bL9vdT".file__is_valid  
    FROM
        "udf_bL9vdT" 
    ORDER BY
        "udf_bL9vdT".sys__id  LIMIT 100000 OFFSET 0 SELECT
            "udf_bL9vdT".sys__id,
            "udf_bL9vdT".sys__rand,
            "udf_bL9vdT".file__source,
            "udf_bL9vdT".file__path,
            "udf_bL9vdT".file__size,
            "udf_bL9vdT".file__version,
            "udf_bL9vdT".file__etag,
            "udf_bL9vdT".file__is_latest,
            "udf_bL9vdT".file__last_modified,
            "udf_bL9vdT".file__location,
            "udf_bL9vdT".file__vtype,
            "udf_bL9vdT".file__is_valid  
        FROM
            "udf_bL9vdT" 
        ORDER BY
            "udf_bL9vdT".sys__id  LIMIT 100000 OFFSET 100000 INSERT 
            INTO
                udf_dvcxog
                (sys__id, file__source, file__path, file__size, file__version, file__etag, file__is_latest, file__last_modified, file__location, file__vtype, file__is_valid) 
            VALUES
                (1, '', 'valid.txt', 10, '', '', 1, '1970-01-01 00:00:00', NULL, '', 1) INSERT 
                INTO
                    udf_dvcxog
                    (sys__id, file__source, file__path, file__size, file__version, file__etag, file__is_latest, file__last_modified, file__location, file__vtype, file__is_valid) 
                VALUES
                    (2, '', 'invalid.txt', 5, '', '', 1, '1970-01-01 00:00:00', NULL, '', 0) SELECT
                        anon_1.file__source,
                        anon_1.file__path,
                        anon_1.file__size,
                        anon_1.file__version,
                        anon_1.file__etag,
                        anon_1.file__is_latest,
                        anon_1.file__last_modified,
                        anon_1.file__location,
                        anon_1.file__vtype,
                        anon_1.file__is_valid  
                    FROM
                        (SELECT
                            anon_2.sys__id AS sys__id,
                            anon_2.sys__rand AS sys__rand,
                            anon_2.file__source AS file__source,
                            anon_2.file__path AS file__path,
                            anon_2.file__size AS file__size,
                            anon_2.file__version AS file__version,
                            anon_2.file__etag AS file__etag,
                            anon_2.file__is_latest AS file__is_latest,
                            anon_2.file__last_modified AS file__last_modified,
                            anon_2.file__location AS file__location,
                            anon_2.file__vtype AS file__vtype,
                            anon_2.file__is_valid AS file__is_valid,
                            udf_dvcxog.file__source AS file__source_1,
                            udf_dvcxog.file__path AS file__path_1,
                            udf_dvcxog.file__size AS file__size_1,
                            udf_dvcxog.file__version AS file__version_1,
                            udf_dvcxog.file__etag AS file__etag_1,
                            udf_dvcxog.file__is_latest AS file__is_latest_1,
                            udf_dvcxog.file__last_modified AS file__last_modified_1,
                            udf_dvcxog.file__location AS file__location_1,
                            udf_dvcxog.file__vtype AS file__vtype_1,
                            udf_dvcxog.file__is_valid AS file__is_valid_1  
                        FROM
                            (SELECT
                                "udf_bL9vdT".sys__id AS sys__id,
                                "udf_bL9vdT".sys__rand AS sys__rand,
                                "udf_bL9vdT".file__source AS file__source,
                                "udf_bL9vdT".file__path AS file__path,
                                "udf_bL9vdT".file__size AS file__size,
                                "udf_bL9vdT".file__version AS file__version,
                                "udf_bL9vdT".file__etag AS file__etag,
                                "udf_bL9vdT".file__is_latest AS file__is_latest,
                                "udf_bL9vdT".file__last_modified AS file__last_modified,
                                "udf_bL9vdT".file__location AS file__location,
                                "udf_bL9vdT".file__vtype AS file__vtype,
                                "udf_bL9vdT".file__is_valid AS file__is_valid  
                            FROM
                                "udf_bL9vdT") AS anon_2 
                        LEFT OUTER JOIN
                            udf_dvcxog 
                                ON udf_dvcxog.sys__id = anon_2.sys__id  LIMIT 20 OFFSET 0
                            ) AS anon_1
```

**Comment by EdwardLi-coder** [comment]
> @EdwardLi-coder to help you fully understand what is going on underneath you can set the `DEBUG_SHOW_SQL_QUERIES=1` environment variable to see the SQL generated by DataChain.
> 
> From the following, it looks like it is not possible to update a field under an existing Feature right now as the result will not be included in the final output:
> 
> ```sql
> SELECT
>         "udf_ZCVabk".sys__id,
>         "udf_ZCVabk".sys__rand,
>         "udf_ZCVabk".file__source,
>         "udf_ZCVabk".file__path,
>         "udf_ZCVabk".file__size,
>         "udf_ZCVabk".file__version,
>         "udf_ZCVabk".file__etag,
>         "udf_ZCVabk".file__is_latest,
>         "udf_ZCVabk".file__last_modified,
>         "udf_ZCVabk".file__location,
>         "udf_ZCVabk".file__vtype,
>         "udf_ZCVabk".file__is_valid  
>     FROM
>         "udf_ZCVabk" 
>     ORDER BY
>         "udf_ZCVabk".sys__id  LIMIT 100000 OFFSET 0;
>     SELECT
>         "udf_ZCVabk".sys__id,
>         "udf_ZCVabk".sys__rand,
>         "udf_ZCVabk".file__source,
>         "udf_ZCVabk".file__path,
>         "udf_ZCVabk".file__size,
>         "udf_ZCVabk".file__version,
>         "udf_ZCVabk".file__etag,
>         "udf_ZCVabk".file__is_latest,
>         "udf_ZCVabk".file__last_modified,
>         "udf_ZCVabk".file__location,
>         "udf_ZCVabk".file__vtype,
>         "udf_ZCVabk".file__is_valid  
>     FROM
>         "udf_ZCVabk" 
>     ORDER BY
>         "udf_ZCVabk".sys__id  LIMIT 100000 OFFSET 100000;
>     INSERT 
>     INTO
>         "udf_gCBFP0"
>         (sys__id, file__source, file__path, file__size, file__version, file__etag, file__is_latest, file__last_modified, file__location, file__vtype, file__is_valid) 
>     VALUES
>         (1, '', 'valid.txt', 10, '', '', 1, '1970-01-01 00:00:00', NULL, '', 1);
>     INSERT 
>     INTO
>         "udf_gCBFP0"
>         (sys__id, file__source, file__path, file__size, file__version, file__etag, file__is_latest, file__last_modified, file__location, file__vtype, file__is_valid) 
>     VALUES
>         (2, '', 'invalid.txt', 5, '', '', 1, '1970-01-01 00:00:00', NULL, '', 0);
>     SELECT
>         anon_1.file__source,
>         anon_1.file__path,
>         anon_1.file__size,
>         anon_1.file__version,
>         anon_1.file__etag,
>         anon_1.file__is_latest,
>         anon_1.file__last_modified,
>         anon_1.file__location,
>         anon_1.file__vtype,
>         anon_1.file__is_valid  
>     FROM
>         (SELECT
>             anon_2.sys__id AS sys__id,
>             anon_2.sys__rand AS sys__rand,
>             anon_2.file__source AS file__source,
>             anon_2.file__path AS file__path,
>             anon_2.file__size AS file__size,
>             anon_2.file__version AS file__version,
>             anon_2.file__etag AS file__etag,
>             anon_2.file__is_latest AS file__is_latest,
>             anon_2.file__last_modified AS file__last_modified,
>             anon_2.file__location AS file__location,
>             anon_2.file__vtype AS file__vtype,
>             anon_2.file__is_valid AS file__is_valid,
>             "udf_gCBFP0".file__source AS file__source_1,
>             "udf_gCBFP0".file__path AS file__path_1,
>             "udf_gCBFP0".file__size AS file__size_1,
>             "udf_gCBFP0".file__version AS file__version_1,
>             "udf_gCBFP0".file__etag AS file__etag_1,
>             "udf_gCBFP0".file__is_latest AS file__is_latest_1,
>             "udf_gCBFP0".file__last_modified AS file__last_modified_1,
>             "udf_gCBFP0".file__location AS file__location_1,
>             "udf_gCBFP0".file__vtype AS file__vtype_1,
>             "udf_gCBFP0".file__is_valid AS file__is_valid_1  
>         FROM
>             (SELECT
>                 "udf_ZCVabk".sys__id AS sys__id,
>                 "udf_ZCVabk".sys__rand AS sys__rand,
>                 "udf_ZCVabk".file__source AS file__source,
>                 "udf_ZCVabk".file__path AS file__path,
>                 "udf_ZCVabk".file__size AS file__size,
>                 "udf_ZCVabk".file__version AS file__version,
>                 "udf_ZCVabk".file__etag AS file__etag,
>                 "udf_ZCVabk".file__is_latest AS file__is_latest,
>                 "udf_ZCVabk".file__last_modified AS file__last_modified,
>                 "udf_ZCVabk".file__location AS file__location,
>                 "udf_ZCVabk".file__vtype AS file__vtype,
>                 "udf_ZCVabk".file__is_valid AS file__is_valid  
>             FROM
>                 "udf_ZCVabk") AS anon_2 
>         LEFT OUTER JOIN
>             "udf_gCBFP0" 
>                 ON "udf_gCBFP0".sys__id = anon_2.sys__id
>             ) AS anon_1
> ```

Ok,Thank you very much .I will try use this environment variable.

**Review by shcheklein — DISMISSED** [review]
Please let's review this comment first https://github.com/iterative/datachain/issues/301#issuecomment-2295540414 (cc @dmpetrov )

File: src/datachain/lib/dc.py [hunk 1]
```diff
@@ -1689,3 +1689,19 @@ def limit(self, n: int) -> "Self":
...
+                raise TypeError(f"Signal '{signal}' is not a File object")
+
+            try:
+                with file.open():
+                    file.is_valid = True
+            except OSError:
+                file.is_valid = False
+            return file
+
+        return self.map(check_file, output={signal: File})
```
Comment (dmpetrov) [inline src/datachain/lib/dc.py:1707]: That's a great starting point! I'll provide summary in the request for changes.

**Review by dmpetrov — CHANGES_REQUESTED** [review]
That's a great starting point!

I'd suggest make a few improvement there:

- [ ] Implement this as a mapper in datachain.file, not in the core DC class
- [ ] We need to populate all file specific structures (for all clouds 😅) (size, version, etag, is_latest, last_modified) to make next command actionable (next command should be able to pull data).
- [ ] Please do not introduce is_valid at this point. We can just keep etag empty/0 as a signal that file was not resolved.

**Comment by EdwardLi-coder** [comment]
> That's a great starting point!
> 
> I'd suggest make a few improvement there:
> 
> * [ ]  Implement this as a mapper in datachain.file, not in the core DC class
> * [ ]  We need to populate all file specific structures (for all clouds 😅) (size, version, etag, is_latest, last_modified) to make next command actionable (next command should be able to pull data).
> * [ ]  Please do not introduce is_valid at this point. We can just keep etag empty/0 as a signal that file was not resolved.

@dmpetrov I have followed the requirements and implemented only the resolve method in this PR. If you think we should also implement resolve_uri in this PR, please let me know. Thank you.

**Comment by dmpetrov** [comment]
> If you think we should also implement resolve_uri in this PR, please let me know. Thank you.

@EdwardLi-coder sorry for a delay.
We need only `resolve_file()` (or just `resolve()`) in this PR. However, the functionality was underspecified. The method should return a resolved file record with all required fields. While `is_valid` can be easely derived from a resolved file record like `dc.mutate(is_valid=Column("file.etag") == "")`

More details in the discussion #301

**Comment by EdwardLi-coder** [comment]
> > If you think we should also implement resolve_uri in this PR, please let me know. Thank you.
> 
> @EdwardLi-coder sorry for a delay. We need only `resolve_file()` (or just `resolve()`) in this PR. However, the functionality was underspecified. The method should return a resolved file record with all required fields. While `is_valid` can be easely derived from a resolved file record like `dc.mutate(is_valid=Column("file.etag") == "")`
> 
> More details in the discussion #301

I have updated my code according to your suggestions. After reading the discussion at https://github.com/iterative/datachain/issues/301, I noticed that you mentioned the possibility of needing two functions. Therefore, I wanted to ask if I should implement the resolve_uri function in this PR as well.

WX20240824-181317@2x

**Comment by EdwardLi-coder** [comment]
@dmpetrov I think the resolve_uri you mentioned should be the method that @volkfox suggested to add in #300.

**Comment by dmpetrov** [comment]
> that @volkfox suggested to add in #300.

Exactly. It's a separate issue and better to implement it as a separate PR.

File: src/datachain/lib/file.py [hunk 2]
```diff
@@ -315,6 +316,35 @@ def get_fs(self):
...
+        try:
+            info = client.info(self.path)
+
+            self.size = info.get("size", 0)
+            self.etag = info.get("etag", "")
+            self.version = info.get("version", "")
+            self.is_latest = info.get("is_latest", True)
+            self.last_modified = info.get("last_modified", datetime.now(timezone.utc))
+        except Exception as e:  # noqa: BLE001
+            self.size = -1
```
Comment (dmpetrov) [inline src/datachain/lib/file.py:339]: Please use defaults from the File field definitions like size=0, is_latest=True.

File: src/datachain/lib/file.py [hunk 3]
```diff
@@ -315,6 +316,35 @@ def get_fs(self):
...
+        """
+        Resolve a File object by checking its existence and updating its metadata.
+
+        Returns:
+            File: The resolved File object with updated metadata.
+        """
+        if self._catalog is None:
+            raise RuntimeError("Cannot resolve file: catalog is not set")
+
+        client = self._catalog.get_client(self.source)
```
Comment (dmpetrov) [inline src/datachain/lib/file.py:329]: what it returns if the protocol is not supported? Is check needed here?

File: tests/unit/lib/test_file.py [hunk 4]
```diff
@@ -323,3 +326,50 @@ def test_read_text(tmp_path, catalog):
...
+        "last_modified": datetime(2023, 1, 1, 12, 0, 0),
+    }
+
+    resolved_file = file.resolve()
+
+    assert resolved_file.size == 1024
+    assert resolved_file.etag == "abc123"
+    assert resolved_file.version == "v1"
+    assert resolved_file.is_latest
+    assert resolved_file.last_modified == datetime(2023, 1, 1, 12, 0, 0)
```
Comment (dmpetrov) [inline tests/unit/lib/test_file.py:361]: I'd try to do this in a single assert like:
```python
assert resolved_file == File(size=1024, etag="abc123", ...)
```

File: tests/unit/lib/test_file.py [hunk 5]
```diff
@@ -323,3 +326,50 @@ def test_read_text(tmp_path, catalog):
...
+
+    mock_client.info.side_effect = Exception("File not found")
```
Comment (dmpetrov) [inline tests/unit/lib/test_file.py:363]: I'd suggest extracting this to a separate test. The same for the missing catalog.

**Review by dmpetrov — APPROVED** [review]
Looks great!

Just a couple of small comments are inline.

One more action is needed: we need a `def resolve(file: File) -> File: ...` as a shortcut for mappers. It should enable scenarios like:

```python
dc = DataChain.from_features(my_files=[File(source="s3://my_bucket", "file1.txt"), File(source="s3://my_bucket", "file2.txt")])

from datachain.file import resolve
dc = dc.map(file = resolve, params="my_files")
content = dc.map(res=lambda file: file.read()).collect()
```

Approved. PLease feel free to merge once the issues are resolved.

**Comment by EdwardLi-coder** [comment]
@dmpetrov I have already updated it. Could you please review it again? Thank you.

**Review by rlamy — CHANGES_REQUESTED** [review]
This needs to be tested with actual (simulated) clouds, i.e. using the `cloud_test_catalog` fixture or similar as in [`tests.func.test_datachain`](https://github.com/iterative/datachain/blob/main/tests/func/test_datachain.py). The existing tests are so heavily mocked that they basically just duplicate the implementation.

Also, (though it makes no difference for using `resolve()` as a mapper) I think it's a bit strange that it modifies the File object in place, I think we should treat them as immutable as much as possible, otherwise users will quickly end up in situations where the object doesn't match what's in the DB.

**Comment by EdwardLi-coder** [comment]
> This needs to be tested with actual (simulated) clouds, i.e. using the `cloud_test_catalog` fixture or similar as in [`tests.func.test_datachain`](https://github.com/iterative/datachain/blob/main/tests/func/test_datachain.py). The existing tests are so heavily mocked that they basically just duplicate the implementation.
> 
> Also, (though it makes no difference for using `resolve()` as a mapper) I think it's a bit strange that it modifies the File object in place, I think we should treat them as immutable as much as possible, otherwise users will quickly end up in situations where the object doesn't match what's in the DB.

I have updated it according to your suggestions.If you have any other suggestions, I will update it accordingly. Thank you.

File: tests/unit/lib/test_file.py [hunk 6]
```diff
@@ -323,3 +326,52 @@ def test_read_text(tmp_path, catalog):
...
     assert file.read_text() == data
+
+
+@pytest.mark.parametrize("cloud_type", ["file", "s3", "gs", "azure"], indirect=True)
+def test_resolve_file(cloud_test_catalog):
+    ctc = cloud_test_catalog
+
+    dc = DataChain.from_storage(ctc.src_uri, catalog=ctc.catalog)
+
+    resolved_files = dc.map(file=resolve).collect("file")
```
Comment (rlamy) [inline tests/unit/lib/test_file.py:337]:
```suggestion
    resolved_files = dc.map(file2=resolve).collect("file2")
```
There seems to be a bug in `.map()` where it doesn't behave correctly if you give it an existing column.

**Review by rlamy — CHANGES_REQUESTED** [review]
Thank you! There are still a few issues though (see comments above) and testing can be improved: 
Ideally, `test_resolve_file()` should check something like `File(source=orig_file.source, path=orig_file.path).resolve() == orig_file` instead of just checking that `orig_file.resolve()` looks reasonable. OTOH, I don't think we need to check the file contents. Also, we should test the case of a non-existing file.

File: tests/unit/lib/test_file.py [hunk 7]
```diff
@@ -323,3 +326,52 @@ def test_read_text(tmp_path, catalog):
     file = File(path=file_name, source=f"file://{tmp_path}")
     file._set_stream(catalog, True)
     assert file.read_text() == data
+
+
+@pytest.mark.parametrize("cloud_type", ["file", "s3", "gs", "azure"], indirect=True)
```
[the remark this answers is not in the document]
Comment (EdwardLi-coder) [reply tests/unit/lib/test_file.py:331]: > You don't need this, that's the default parametrisation for `cloud_type`.

Hi @rlamy.I understand that you meant I should remove ["file", "s3", "gs", "azure"]. I have removed them, but an error has occurred:

```
tests/unit/lib/test_file.py:332:2: error: Missing positional argument "argvalues" in call to "__call__" of "_ParametrizeMarkDecorator" [call-arg]
    @pytest.mark.parametrize("cloud_type", indirect=True)
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tests/unit/lib/test_file.py:346:2: error: Missing positional argument "argvalues" in call to "__call__" of "_ParametrizeMarkDecorator" [call-arg]
    @pytest.mark.parametrize("cloud_type", indirect=True)
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Found 2 errors in 1 file (checked 181 source files)
```

I'm not sure if I understood your instructions correctly. If not, please let me know. Thank you.

[the remark this answers is not in the document]
Comment (rlamy) [reply tests/unit/lib/test_file.py:331]: I meant that you can remove the whole decorator.

**Comment by EdwardLi-coder** [comment]
> Thank you! There are still a few issues though (see comments below) and testing can be improved: Ideally, `test_resolve_file()` should check something like `File(source=orig_file.source, path=orig_file.path).resolve() == orig_file` instead of just checking that `orig_file.resolve()` looks reasonable. OTOH, I don't think we need to check the file contents. Also, we should test the case of a non-existing file.

I updated it.

File: src/datachain/lib/file.py [hunk 8]
```diff
@@ -316,6 +319,72 @@ def get_fs(self):
         """Returns `fsspec` filesystem for the file."""
         return self._catalog.get_client(self.source).fs
 
+    def resolve(self) -> "File":
```
Comment (rlamy) [inline src/datachain/lib/file.py:322]:
```suggestion
    def resolve(self) -> "Self":
```

File: src/datachain/lib/file.py [hunk 9]
```diff
@@ -316,6 +319,72 @@ def get_fs(self):
...
+        try:
+            client = self._catalog.get_client(self.source)
+        except NotImplementedError as e:
+            raise RuntimeError(
+                f"Unsupported protocol for file source: {self.source}"
+            ) from e
+
+        try:
+            info = client.fs.info(client.get_full_path(self.path))
+            converted_info = client.convert_info(info, self.source)
```
Comment (rlamy) [inline src/datachain/lib/file.py:341]: We're actually removing `.convert_info()` (sorry!), please use [`client.info_to_file()`](https://github.com/iterative/datachain/blob/43f5b21f70581f76fa26732b8234af3b6f76c0ea/src/datachain/client/azure.py#L28) instead. Also, that way all attributes will be present and have a correct value.

**Review by rlamy — COMMENTED** [review]
Sorry for the late review! This looks good, except for replacing `.convert_info()` with `.info_to_file()`.

**Comment by EdwardLi-coder** [comment]
> Sorry for the late review! This looks good, except for replacing `.convert_info()` with `.info_to_file()`.

I updated it.

File: src/datachain/lib/file.py [hunk 10]
```diff
@@ -313,6 +319,70 @@ def get_fs(self):
...
+
+        try:
+            info = client.fs.info(client.get_full_path(self.path))
+            converted_info = client.info_to_file(info, self.source)
+            return type(self)(
+                path=self.path,
+                source=self.source,
+                size=converted_info.size,
+                etag=converted_info.etag,
+                version=converted_info.version or "",
```
Comment (rlamy) [inline src/datachain/lib/file.py:347]: You shouldn't need this `or ""` (since https://github.com/iterative/datachain/pull/419 was merged)

**Review by rlamy — COMMENTED** [review]
Thank you! One last rebase and I think it can be merged.

**Comment by EdwardLi-coder** [comment]
> Thank you! One last rebase and I think it can be merged.

I updated it.Thank you.

**Review by rlamy — APPROVED** [review]
LGTM, thanks!

**Pull request merged**